Virtual Clients for RM_Call - enables  modules to use WATCH

sjpotter opened the pull request.

sjpotter committed:
RedisModuleClient - enables MULTI/EXEC/WATCH in RM_Call()

This adds a new object the RedisModuleClient that modules can allocate a client that can persistent between many usage of RM_Call().

This enables modules to use Watch/Multi/Exec to benefit from Redis's conception of transactions.

It adds 4 new RM_ APIs

- RM_CreateModuleClient() - creates the object.
- RM_FreeModuleClient() - frees the object.
- RM_SetContextClient() - associates object with a given context so that RM_Call will use it instead of a temp client.
- RM_GetClientFlags() - gets the flag set that the RM API exposes (not all flags, currently just the 2 dirty flags).

tezc commented on src/module.c:
```diff
@@ -5981,10 +5989,52 @@ RedisModuleString *RM_CreateStringFromCallReply(RedisModuleCallReply *reply) {
     ctx->user = user;
 }
 
+RedisModuleClient *RM_CreatePersistentClient(RedisModuleCtx *ctx, const RedisModuleUser *rmu)
```
Btw persistent client sounds like something that we will store in rdb. (as if like something to be written to the disk). Personal opinion, "moduleclient" makes more sense to me, like RM_CreateModuleClient() ..

sjpotter replied:
ok, changed.

tezc commented on src/module.c:
```diff
@@ -5981,10 +5989,52 @@ RedisModuleString *RM_CreateStringFromCallReply(RedisModuleCallReply *reply) {
+void RM_ClearClientDirty(RedisModuleClient *client)
+{
+    client->client->flags &= ~CLIENT_DIRTY_CAS;
+}
```
I remember there was a discussion about how to serialize/deserialize client object. So, if we add more fields (say more flags), we will need to introduce new APIs to get/restore those flags. If we make it opaque, it might be more flexible. Was there a conclusion in that discussion? Do you remember it?

sjpotter replied:
I don't think it should be opaque.  At best it should be RM_GetClientFlags() and it return a set of redismodule.h defined flags.

The Q is, should one be able to reset ClientDirty()?  perhaps not, and one would just free/reallocate a new client.

sjpotter committed:
add RM_SetClientUser()

sjpotter committed:
user doesn't have to be part of RM_CreateModuleClient anymore

sjpotter committed:
acl for module clients + fix header

sjpotter committed:
improve handling of releasing / resetting persistent client

sjpotter committed:
typo

sjpotter committed:
refactor module releasing to support blocked calls

added a flag to client to know if its owned by a module directly.  changed how the clients are released based on this to allow it to work in context of blocking calls

addressed most of oran's comments

sjpotter committed:
tests for blocking commands with persistent clients

currently the tests fail inthe tcl code, but work when I run by hand, need to figure out what I'm doing wrong

sjpotter committed:
fix tcl tests

sjpotter committed:
blpop in multi should still use allow blocking flag

sjpotter committed:
Update src/server.c

Co-authored-by: Oran Agra <oran@redislabs.com>

sjpotter committed:
revert order of persistent client user and context user usage + tests for it

sjpotter committed:
Merge remote-tracking branch 'origin/rm-call-persistent-client' into rm-call-persistent-client

sjpotter committed:
last comment can address right now

sjpotter committed:
improve resetting of client flags on end of execution unit

sjpotter committed:
lint issue for pubsub unsubscribe functions

sjpotter committed:
fix reset_flags, needs to be set to 0 upfront and reset every time used

the reset_flag had junk data in it previously it seems causing it to reset the fact that this is flagged as persistent client.

now set it to 0 at creation and reset it to 0 every time we reset the flags (as shouldn't carry through to next exectution).

sjpotter committed:
move reset_flags to bottom of struct

enjoy-binbin committed:
Optimize HRANDFIELD and ZRANDMEMBER case 3 when listpack encoded (#12205)

Optimized HRANDFIELD and ZRANDMEMBER commands as in #8444,
CASE 3 under listpack encoding. Boost optimization to CASE 2.5. 

CASE 2.5 listpack only. Sampling unique elements, in non-random order.
Listpack encoded hashes / zsets are meant to be relatively small, so
HRANDFIELD_SUB_STRATEGY_MUL / ZRANDMEMBER_SUB_STRATEGY_MUL
isn't necessary and we rather not make copies of the entries. Instead, we
emit them directly to the output buffer.

Simple benchmarks shows it provides some 400% improvement in HRANDFIELD
and ZRANGESTORE both in CASE 3.

Unrelated changes: remove useless setTypeRandomElements and fix a typo.

sjpotter committed:
Fix memory leak when RM_Call's RUN_AS_USER fails (#12158)

previously the argv wasn't freed so would leak.  not a common case, but should be handled.

Solution: move RUN_AS_USER setup and error exit to the right place.
this way, when we do `goto cleanup` (instead of return) it'll automatically do the right thing (including autoMemoryAdd)
Removed the user argument from moduleAllocTempClient (reverted to the state before 6e993a5)

Co-authored-by: Oran Agra <oran@redislabs.com>

sjpotter committed:
Merge remote-tracking branch 'upstream/unstable' into rm-call-persistent-client
